// @goodsdks/citizen-sdk/dist/index.js
import { parseAbi, zeroAddress, ContractFunctionExecutionError, createPublicClient, http } from 'viem';
import { waitForTransactionReceipt } from 'viem/actions';
import { compressToEncodedURIComponent } from 'lz-string';

// src/sdks/viem-identity-sdk.ts
var FV_IDENTIFIER_MSG2 = `Sign this message to request verifying your account <account> and to create your own secret unique identifier for your anonymized record.
You can use this identifier in the future to delete this anonymized record.
WARNING: do not sign this message unless you trust the website/application requesting this signature.`;
var Envs = {
  production: {
    dappUrl: "https://wallet.gooddollar.org",
    identityUrl: "https://goodid.gooddollar.org",
    backend: "https://goodserver.gooddollar.org",
    goodCollectiveUrl: "https://goodcollective.vercel.app/"
  },
  staging: {
    dappUrl: "https://qa.gooddollar.org",
    identityUrl: "https://goodid-qa.vercel.app",
    backend: "https://goodserver-qa.herokuapp.com",
    goodCollectiveUrl: "https://dev-goodcollective.vercel.app/"
  },
  development: {
    dappUrl: "https://dev.gooddollar.org",
    identityUrl: "https://goodid-dev.vercel.app",
    backend: "https://good-server.herokuapp.com",
    goodCollectiveUrl: "https://dev-goodcollective.vercel.app/"
  }
};
var SupportedChains = /* @__PURE__ */ ((SupportedChains6) => {
  SupportedChains6[SupportedChains6["FUSE"] = 122] = "FUSE";
  SupportedChains6[SupportedChains6["CELO"] = 42220] = "CELO";
  SupportedChains6[SupportedChains6["XDC"] = 50] = "XDC";
  return SupportedChains6;
})(SupportedChains || {});
var FALLBACK_CHAIN_PRIORITY = [
  42220 /* CELO */,
  50 /* XDC */,
  122 /* FUSE */
];
var CHAIN_DECIMALS = {
  [122 /* FUSE */]: 2,
  [42220 /* CELO */]: 18,
  [50 /* XDC */]: 18
};
var makeExplorer = (baseUrl) => ({
  address: (address) => `${baseUrl}/address/${address}`,
  tx: (hash) => `${baseUrl}/tx/${hash}`
});
var chainConfigs = {
  [122 /* FUSE */]: {
    id: 122 /* FUSE */,
    key: "FUSE",
    label: "Fuse",
    shortName: "Fuse",
    explorer: makeExplorer("https://explorer.fuse.io"),
    rpcUrls: ["https://rpc.fuse.io", "https://fuse-rpc.gateway.pokt.network"],
    defaultGasPrice: BigInt(11e9),
    claimGasBuffer: 150000n,
    fvDefaultChain: 42220 /* CELO */,
    contracts: {
      production: {
        identityContract: "0x2F9C28de9e6d44b71B91b8BA337A5D82e308E7BE",
        ubiContract: "0xd253A5203817225e9768C05E5996d642fb96bA86",
        faucetContract: "0x01ab5966C1d742Ae0CFF7f14cC0F4D85156e83d9",
        g$Contract: "0x495d133B938596C9984d462F007B676bDc57eCEC"
      },
      staging: {
        identityContract: "0xb0cD4828Cc90C5BC28f4920Adf2Fd8F025003D7E",
        ubiContract: "0x54469071Ca82B46A2C01C09D38ca6Ca4347EB21d",
        faucetContract: "0x70f361EDB97B245E8A68573637A31886A427fe2a",
        g$Contract: "0xe39236a9Cf13f65DB8adD06BD4b834C65c523d2b"
      },
      development: {
        identityContract: "0x1e006225cff7d37411db28f652e0Da9D20325eBb",
        ubiContract: "0x3bdeB796950301FfC9568fAF89B7370f8B217321",
        faucetContract: "0x09Ad3430D146aa662eA8c20cBEBceBC0bbB3FB0a",
        g$Contract: "0x79BeecC4b165Ccf547662cB4f7C0e83b3796E5b3"
      }
    }
  },
  [42220 /* CELO */]: {
    id: 42220 /* CELO */,
    key: "CELO",
    label: "Celo",
    shortName: "Celo",
    explorer: makeExplorer("https://celoscan.io"),
    rpcUrls: ["https://forno.celo.org", "https://rpc.ankr.com/celo"],
    defaultGasPrice: BigInt(25001e6),
    claimGasBuffer: 250000n,
    fvDefaultChain: 42220 /* CELO */,
    contracts: {
      production: {
        identityContract: "0xC361A6E67822a0EDc17D899227dd9FC50BD62F42",
        ubiContract: "0x43d72Ff17701B2DA814620735C39C620Ce0ea4A1",
        faucetContract: "0x4F93Fa058b03953C851eFaA2e4FC5C34afDFAb84",
        g$Contract: "0x62B8B11039FcfE5aB0C56E502b1C372A3d2a9c7A"
      },
      staging: {
        identityContract: "0x0108BBc09772973aC27983Fc17c7D82D8e87ef4D",
        ubiContract: "0x2881d417dA066600372753E73A3570F0781f18cB",
        faucetContract: "0x9A0F8AEc626A0f493941Ceb1dA6021cFB0567293",
        g$Contract: "0x61FA0fB802fd8345C06da558240E0651886fec69"
      },
      development: {
        identityContract: "0xF25fA0D4896271228193E782831F6f3CFCcF169C",
        ubiContract: "0x6B86F82293552C3B9FE380FC038A89e0328C7C5f",
        faucetContract: "0x635b420e95b364def3A031166dA4bC4F57bf9dEB",
        g$Contract: "0xFa51eFDc0910CCdA91732e6806912Fa12e2FD475"
      }
    }
  },
  [50 /* XDC */]: {
    id: 50 /* XDC */,
    key: "XDC",
    label: "XDC Network",
    shortName: "XDC",
    explorer: makeExplorer("https://xdcscan.com"),
    rpcUrls: ["https://rpc.ankr.com/xdc"],
    defaultGasPrice: BigInt(125e8),
    claimGasBuffer: 150000n,
    fvDefaultChain: 50 /* XDC */,
    contracts: {
      production: {
        identityContract: "0x27a4a02C9ed591E1a86e2e5D05870292c34622C9",
        ubiContract: "0x22867567E2D80f2049200E25C6F31CB6Ec2F0faf",
        faucetContract: "0x7344Da1Be296f03fbb8082aDaC5696058B5a9bd9",
        g$Contract: "0xEC2136843a983885AebF2feB3931F73A8eBEe50c"
      },
      development: {
        identityContract: "0xa6632e9551A340E8582cc797017fbA645695E29f",
        ubiContract: "0xA2619D468EfE2f6D8b3D915B999327C8fE13aE2c",
        faucetContract: "0x5CE5e22329914F3317B770184a405E91ef72e489",
        g$Contract: "0xA13625A72Aef90645CfCe34e25c114629d7855e7"
      }
    }
  }
};
var chainIds = Object.values(SupportedChains).filter(
  (value) => typeof value === "number"
);
var chainIdSet = new Set(chainIds);
var SUPPORTED_CHAIN_IDS = chainIds;
var isSupportedChain = (chainId) => typeof chainId === "number" && chainIdSet.has(chainId);
var createRpcUrlIterator = (chainId) => {
  const urls = [...chainConfigs[chainId]?.rpcUrls ?? []];
  if (!urls.length) {
    throw new Error(`No RPC URLs configured for chain ${chainId}`);
  }
  let index = 0;
  return () => {
    const url = urls[index];
    index = (index + 1) % urls.length;
    return url;
  };
};
var identityV2ABI = parseAbi([
  // Whitelist management (IdentityV2)
  "function addWhitelisted(address account)",
  "function removeWhitelisted(address account)",
  "function getWhitelistedRoot(address account) view returns (address)",
  "function lastAuthenticated(address account) view returns (uint256)",
  "function authenticationPeriod() view returns (uint256)",
  // Wallet-link methods (IdentityV4)
  "function connectAccount(address account) external",
  "function disconnectAccount(address account) external",
  "function connectedAccounts(address account) view returns (address)"
]);
var ubiSchemeV2ABI = parseAbi([
  "function claim() returns (bool)",
  "function checkEntitlement() view returns (uint256)",
  "function checkEntitlement(address _member) view returns (uint256)",
  "function getDailyStats() view returns (uint256 claimers, uint256 amount)",
  "function periodStart() view returns (uint256)",
  "function currentDay() view returns (uint256)",
  "event UBIClaimed(address indexed account, uint256 amount)"
]);
var faucetABI = parseAbi([
  "function minTopping() view returns (uint8)",
  "function getToppingAmount() view returns (uint256)",
  "function canTop(address _user) view returns (bool)",
  "function topWallet(address _user) payable"
]);
var g$ABI = parseAbi([
  "function balanceOf(address account) view returns (uint256)"
]);
var WALLET_LINK_SECURITY_MESSAGES = {
  connect: [
    "SECURITY NOTICE \u2014 Connect Wallet",
    "You are about to link a secondary wallet to your GoodDollar identity.",
    "\u2022 Only connect wallets you own and control.",
    "\u2022 All connected wallets share a single daily UBI claim.",
    "\u2022 The whitelisted root identity is the only account that can connect a wallet.",
    "Do not proceed if you did not initiate this action."
  ].join("\n"),
  disconnect: [
    "SECURITY NOTICE \u2014 Disconnect Wallet",
    "You are about to remove a wallet from your GoodDollar identity.",
    "\u2022 Once disconnected the wallet no longer shares your UBI claim.",
    "\u2022 Either the root identity or the connected account itself can disconnect.",
    "Do not proceed if you did not initiate this action."
  ].join("\n")
};

// src/utils/chains.ts
var resolveChainAndContract = (walletClient, env) => {
  const chainId = walletClient.chain?.id;
  if (!isSupportedChain(chainId)) {
    throw new Error(`Unsupported chain ID.`);
  }
  const contractEnvAddresses = chainConfigs[chainId]?.contracts[env] ?? null;
  if (!contractEnvAddresses) {
    throw new Error(
      `Contract address for environment "${env}" on chain "${chainConfigs[chainId].label}" not found. Try a different environment or chain.`
    );
  }
  return { chainId, contractEnvAddresses };
};
var createRpcIteratorRegistry = () => /* @__PURE__ */ new Map();
var getIterator = (chainId, registry) => {
  if (!registry.has(chainId)) {
    registry.set(chainId, createRpcUrlIterator(chainId));
  }
  const iterator = registry.get(chainId);
  if (!iterator) {
    throw new Error(`RPC iterator not available for chain ${chainId}`);
  }
  return iterator;
};
var getRpcFallbackClient = (chainId, registry) => {
  const iterator = getIterator(chainId, registry);
  const rpcUrl = iterator();
  return createPublicClient({
    transport: http(rpcUrl)
  });
};
var shouldRetryRpcFallback = (errorMessage, chainId, attempt) => {
  if (attempt > 0) {
    return false;
  }
  const rpcUrls = chainConfigs[chainId]?.rpcUrls ?? [];
  if (!rpcUrls.length) {
    return false;
  }
  const normalizedMessage = errorMessage.toLowerCase();
  return normalizedMessage.includes("transports[i] is not a function");
};
var extractErrorMessage = (error) => {
  const messages = /* @__PURE__ */ new Set();
  const err = error;
  if (typeof err?.shortMessage === "string") {
    messages.add(err.shortMessage);
  }
  if (typeof err?.message === "string") {
    messages.add(err.message);
  }
  if (typeof err?.details === "string") {
    messages.add(err.details);
  }
  const causeMessage = err?.cause?.message;
  if (typeof causeMessage === "string") {
    messages.add(causeMessage);
  }
  if (!messages.size) {
    return "Unknown error";
  }
  return Array.from(messages).join(" | ");
};

// src/sdks/viem-identity-sdk.ts
var initializeIdentityContract = (publicClient, contractAddress) => ({
  publicClient,
  contractAddress
});
var IdentitySDK = class _IdentitySDK {
  account;
  publicClient;
  walletClient;
  contract;
  env = "production";
  chainId;
  fvDefaultChain;
  rpcIterators = createRpcIteratorRegistry();
  /**
   * Initializes the IdentitySDK.
   * @param publicClient - The PublicClient instance.
   * @param walletClient - The WalletClient with WalletActions.
   * @param env - The environment to use ("production" | "staging" | "development").
   */
  constructor({
    account,
    publicClient,
    walletClient,
    env
  }) {
    if (!walletClient.account) {
      throw new Error(
        "IdentitySDK: WalletClient must have an account attached."
      );
    }
    this.publicClient = publicClient;
    this.walletClient = walletClient;
    this.env = env;
    this.account = account ?? walletClient.account.address;
    const { chainId, contractEnvAddresses } = resolveChainAndContract(
      walletClient,
      env
    );
    this.chainId = chainId;
    this.fvDefaultChain = chainConfigs[chainId]?.fvDefaultChain ?? chainId;
    this.contract = initializeIdentityContract(
      this.publicClient,
      contractEnvAddresses.identityContract
    );
  }
  static async init(props) {
    const [account] = await props.walletClient.getAddresses();
    return new _IdentitySDK({ account, ...props });
  }
  /**
   * Evaluates security constraints and prompts the user with security messages if required.
   * Custodial/automated flows can bypass this entirely via `skipSecurityMessage: true`.
   * For native wallet flows, `onSecurityMessage` receives the notice and must resolve
   * to `true` for the transaction to proceed. Without either option the message is
   * logged to `console.info` as a headless default.
   * @param action - The wallet link action being performed ("connect" or "disconnect").
   * @param options - Additional wallet link options including security prompt handlers.
   */
  async runSecurityCheck(action, options) {
    if (options?.skipSecurityMessage) return;
    const message = WALLET_LINK_SECURITY_MESSAGES[action];
    if (options?.onSecurityMessage) {
      const confirmed = await options.onSecurityMessage(message);
      if (!confirmed) {
        throw new Error(
          `Wallet ${action} cancelled: user did not confirm security notice.`
        );
      }
      return;
    }
    console.info(`[IdentitySDK] ${message}`);
  }
  /**
   * Submits a transaction and waits for its receipt.
   * @param params - Parameters for simulating the contract call.
   * @param onHash - Optional callback to receive the transaction hash.
   * @returns The transaction receipt.
   * @throws If submission fails or no active wallet address is found.
   */
  async submitAndWait(params, onHash) {
    try {
      if (!this.account) throw new Error("No active wallet address found.");
      const { request } = await this.publicClient.simulateContract({
        account: this.account,
        ...params
      });
      const hash = await this.walletClient.writeContract(request);
      onHash?.(hash);
      return waitForTransactionReceipt(this.publicClient, { hash });
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      console.error("submitAndWait Error:", error);
      throw new Error(`Failed to submit transaction: ${message}`);
    }
  }
  /**
   * Returns whitelist status of main account or any connected account.
   * @param account - The account address to check.
   * @returns An object containing whitelist status and root address.
   * @reference: https://docs.gooddollar.org/user-guides/connect-another-wallet-address-to-identity
   */
  async getWhitelistedRoot(account) {
    try {
      const root = await this.publicClient.readContract({
        address: this.contract.contractAddress,
        abi: identityV2ABI,
        functionName: "getWhitelistedRoot",
        args: [account]
      });
      return {
        isWhitelisted: root !== zeroAddress,
        root
      };
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      console.error("getWhitelistedRoot Error:", error);
      throw new Error(`Failed to get whitelisted root: ${message}`);
    }
  }
  /**
   * Checks the wallet-link connection status of the given account.
   *
   * When `chainId` is provided the query is scoped to that single chain only.
   * When omitted, all supported chains are queried in parallel via Promise.allSettled
   * and one ChainConnectedStatus entry is returned per chain.
   *
   * App-provided public clients always take precedence. For the SDK's current
   * chain `this.publicClient` is reused when no override is passed. For other
   * chains the SDK falls back to local read-only RPC clients using the same
   * fallback flow as other multi-chain reads.
   *
   * @param account - The account address to check.
   * @param chainId - Optional. Restricts the query to this chain only.
   * @param publicClients - Optional. App-level public clients keyed by SupportedChains ID.
   * @returns An array of ChainConnectedStatus (one entry per queried chain).
   */
  async checkConnectedStatus(account, chainId, publicClients) {
    const configs = chainId ? Object.values(chainConfigs).filter((config) => config.id === chainId) : Object.values(chainConfigs);
    const settled = await Promise.allSettled(
      configs.map(async (config) => {
        const contracts = config.contracts[this.env];
        if (!contracts) {
          return {
            chainId: config.id,
            chainName: config.label,
            isConnected: false,
            root: zeroAddress,
            error: `No contract configured for env "${this.env}" on ${config.label}`
          };
        }
        const isPrimaryChain = this.chainId === config.id;
        const client = publicClients?.[config.id] ?? (isPrimaryChain ? this.publicClient : getRpcFallbackClient(config.id, this.rpcIterators));
        const root = await client.readContract({
          address: contracts.identityContract,
          abi: identityV2ABI,
          functionName: "connectedAccounts",
          args: [account]
        });
        return {
          chainId: config.id,
          chainName: config.label,
          isConnected: root !== zeroAddress,
          root
        };
      })
    );
    return settled.map((result, i) => {
      const config = configs[i];
      if (result.status === "fulfilled") return result.value;
      return {
        chainId: config.id,
        chainName: config.label,
        isConnected: false,
        root: zeroAddress,
        error: result.reason?.message ?? "Unknown RPC error"
      };
    });
  }
  /**
   * Retrieves identity expiry data for a given account.
   * @param account - The account address.
   * @returns The identity expiry data.
   */
  async getIdentityExpiryData(account) {
    try {
      const [lastAuthenticated, authPeriod] = await Promise.all([
        this.publicClient.readContract({
          address: this.contract.contractAddress,
          abi: identityV2ABI,
          functionName: "lastAuthenticated",
          args: [account]
        }),
        this.publicClient.readContract({
          address: this.contract.contractAddress,
          abi: identityV2ABI,
          functionName: "authenticationPeriod",
          args: []
        })
      ]);
      return { lastAuthenticated, authPeriod };
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      console.error("getIdentityExpiryData Error:", error);
      throw new Error(
        `Failed to retrieve identity expiry data: ${message}`
      );
    }
  }
  /**
   * Updates the wallet-link status of an account.
   * The whitelisted root identity must be the signer for `connectAccount`,
   * while `disconnectAccount` can be called by either the root identity or
   * the connected account itself.
   * Custodial flows can pass `skipSecurityMessage: true` to bypass the notice;
   * the underlying transaction signing is handled by the app-provided walletClient
   * (or IdentityCustodialSDK for LocalAccount signers).
   * @param action - The wallet-link action to submit.
   * @param account - The account address to update.
   * @param options - Additional options, such as security prompts.
   * @returns The transaction receipt.
   */
  async updateConnectionStatus(action, account, options) {
    const securityAction = action === "connectAccount" ? "connect" : "disconnect";
    const actionLabel = action === "connectAccount" ? "connect" : "disconnect";
    try {
      await this.runSecurityCheck(securityAction, options);
      return this.submitAndWait(
        {
          address: this.contract.contractAddress,
          abi: identityV2ABI,
          functionName: action,
          args: [account]
        },
        options?.onHash
      );
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      console.error("updateConnectionStatus Error:", error);
      throw new Error(`Failed to ${actionLabel} account: ${message}`);
    }
  }
  /**
   * Connects a new account to the identity.
   * The whitelisted root identity must be the signer. No additional message
   * signature is required beyond the transaction itself.
   * @param account - The account address to connect.
   * @param options - Additional options, such as security prompts.
   * @returns The transaction receipt.
   */
  async connectAccount(account, options) {
    return this.updateConnectionStatus("connectAccount", account, options);
  }
  /**
   * Disconnects an account from the identity.
   * Either the root identity or the connected account itself can call this.
   * @param account - The account address to disconnect.
   * @param options - Additional options, such as security prompts.
   * @returns The transaction receipt.
   */
  async disconnectAccount(account, options) {
    return this.updateConnectionStatus("disconnectAccount", account, options);
  }
  /**
   * Generates a Face Verification Link.
   * @param popupMode - Whether to generate a popup link.
   * @param callbackUrl - The URL to callback after verification.
   * @param chainId - The blockchain network ID.
   * @returns The generated Face Verification link.
   */
  async generateFVLink(popupMode = false, callbackUrl, chainId) {
    try {
      const address = this.account;
      if (!address) throw new Error("No wallet address found.");
      const nonce = Math.floor(Date.now() / 1e3).toString();
      const fvSigMessage = FV_IDENTIFIER_MSG2.replace("<account>", address);
      const fvSig = await this.walletClient.signMessage({
        account: address,
        message: fvSigMessage
      });
      const { identityUrl } = Envs[this.env];
      if (!identityUrl) {
        throw new Error("identityUrl is not defined in environment settings.");
      }
      if (!fvSig) {
        throw new Error("Missing signature for Face Verification.");
      }
      if (!popupMode && !callbackUrl) {
        throw new Error("Callback URL is required for redirect mode.");
      }
      const url = new URL(identityUrl);
      const fallbackChain = this.fvDefaultChain ?? this.chainId;
      const resolvedChain = typeof chainId === "number" ? chainId : fallbackChain;
      const fvChain = isSupportedChain(resolvedChain) ? resolvedChain : fallbackChain;
      const params = {
        account: address,
        nonce,
        fvsig: fvSig,
        chain: fvChain
      };
      if (callbackUrl) {
        params[popupMode ? "cbu" : "rdu"] = callbackUrl;
      }
      url.searchParams.append(
        "lz",
        compressToEncodedURIComponent(JSON.stringify(params))
      );
      return url.toString();
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      console.error("generateFVLink Error:", error);
      throw new Error(
        `Failed to generate Face Verification link: ${message}`
      );
    }
  }
  /**
   * Calculates the identity expiry timestamp.
   * @param lastAuthenticated - The timestamp of last authentication.
   * @param authPeriod - The authentication period.
   * @returns The identity expiry data.
   */
  calculateIdentityExpiry(lastAuthenticated, authPeriod) {
    const MS_IN_A_SECOND = 1e3;
    const MS_IN_A_DAY = 24 * 60 * 60 * MS_IN_A_SECOND;
    const periodInMs = authPeriod * BigInt(MS_IN_A_DAY);
    const expiryTimestamp = lastAuthenticated * BigInt(MS_IN_A_SECOND) + periodInMs;
    return { expiryTimestamp };
  }
};
async function canClaim(chainId, account, publicClient, faucetAddress, toppingAmount, minTopping) {
  const fees = await publicClient.estimateFeesPerGas().catch((error) => console.log("Error estimating fees per gas:", error));
  const gasPrice = fees?.maxFeePerGas ?? fees?.gasPrice ?? chainConfigs[chainId]?.defaultGasPrice;
  if (!gasPrice) {
    throw new Error(
      "Cannot determine gasPrice for the current connected chain."
    );
  }
  const claimGasBuffer = chainConfigs[chainId]?.claimGasBuffer ?? void 0;
  if (claimGasBuffer == null) {
    throw new Error(
      "Cannot determine gas buffer requirement for the current connected chain."
    );
  }
  const minBalance = claimGasBuffer * gasPrice;
  const minThreshold = toppingAmount * (100n - BigInt(minTopping)) / 100n || minBalance;
  const balance = await publicClient.getBalance({
    address: account
  });
  return balance >= minThreshold;
}
async function triggerFaucet({
  chainId,
  account,
  publicClient,
  walletClient,
  faucetAddress,
  env,
  throttleMs = 60 * 60 * 1e3
}) {
  if (typeof localStorage !== "undefined") {
    const key = `goodDollarFaucetLastToppedUtcMs_${chainId}`;
    const last = localStorage.getItem(key);
    if (last && Date.now() < Number(last) + throttleMs) {
      return "skipped";
    }
  }
  try {
    const [balance, canTop, toppingAmount, minTopping] = await Promise.all([
      publicClient.getBalance({ address: account }),
      publicClient.readContract({
        address: faucetAddress,
        abi: faucetABI,
        functionName: "canTop",
        args: [account],
        account
      }),
      publicClient.readContract({
        address: faucetAddress,
        abi: faucetABI,
        functionName: "getToppingAmount",
        args: [],
        account
      }),
      publicClient.readContract({
        address: faucetAddress,
        abi: faucetABI,
        functionName: "minTopping",
        args: [],
        account
      })
    ]);
    const hasGoodBalance = await canClaim(
      chainId,
      account,
      publicClient,
      faucetAddress,
      toppingAmount,
      minTopping
    );
    if (!canTop || balance >= toppingAmount || hasGoodBalance) {
      if (typeof localStorage !== "undefined") {
        localStorage.setItem(
          `goodDollarFaucetLastToppedUtcMs_${chainId}`,
          String(Date.now())
        );
      }
      return "skipped";
    }
    const { request } = await publicClient.simulateContract({
      address: faucetAddress,
      abi: faucetABI,
      functionName: "topWallet",
      args: [account],
      account,
      chain: walletClient.chain
    });
    const estimatedGas = await publicClient.estimateContractGas({
      address: faucetAddress,
      abi: faucetABI,
      functionName: "topWallet",
      args: [account],
      account
    });
    const fees = await publicClient.estimateFeesPerGas().catch(() => void 0);
    const perGasWei = fees?.maxFeePerGas ?? fees?.gasPrice ?? chainConfigs[chainId]?.defaultGasPrice;
    if (typeof perGasWei !== "bigint") {
      throw new Error("Unable to determine per-gas price for faucet top-up");
    }
    const estimatedTxFeeWei = estimatedGas * perGasWei;
    if (balance < estimatedTxFeeWei) {
      throw new Error("Not enough balance to pay for estimated gas fee");
    }
    if (estimatedTxFeeWei > toppingAmount) {
      throw new Error("Estimated gas fee exceeds topping amount");
    }
    const hash = await walletClient.writeContract(request);
    await new Promise((r) => setTimeout(r, 2e3));
    await waitForTransactionReceipt(publicClient, {
      hash,
      retryDelay: 3e3
    });
    if (typeof localStorage !== "undefined") {
      localStorage.setItem(
        `goodDollarFaucetLastToppedUtcMs_${chainId}`,
        String(Date.now())
      );
    }
    return "topped_via_contract";
  } catch (err) {
    console.error("Faucet topWallet error, falling back to API", err);
    try {
      const { backend } = Envs[env] || {};
      if (!backend) return "error";
      const response = await fetch(`${backend}/verify/topWallet`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ chainId, account })
      });
      if (!response.ok) return "error";
      if (typeof localStorage !== "undefined") {
        localStorage.setItem(
          `goodDollarFaucetLastToppedUtcMs_${chainId}`,
          String(Date.now())
        );
      }
      return "topped_via_api";
    } catch {
      return "error";
    }
  }
}

// src/sdks/viem-claim-sdk.ts
var DAY = 1e3 * 60 * 60 * 24;
async function checkGenericEntitlement(options) {
  const env = options.env ?? "production";
  const config = chainConfigs[options.chainId];
  const contracts = config?.contracts?.[env];
  if (!config || !contracts) {
    throw new Error(
      `No UBIScheme contract configuration for chain ${options.chainId} in env "${env}"`
    );
  }
  return options.publicClient.readContract({
    address: contracts.ubiContract,
    abi: ubiSchemeV2ABI,
    functionName: "checkEntitlement",
    args: []
  });
}
async function checkGenericDailyStats(options) {
  const env = options.env ?? "production";
  const config = chainConfigs[options.chainId];
  const contracts = config?.contracts?.[env];
  if (!config || !contracts) {
    throw new Error(
      `No UBIScheme contract configuration for chain ${options.chainId} in env "${env}"`
    );
  }
  const [claimers, amount] = await options.publicClient.readContract({
    address: contracts.ubiContract,
    abi: ubiSchemeV2ABI,
    functionName: "getDailyStats",
    args: []
  });
  return { claimers, amount };
}
var ClaimSDK = class _ClaimSDK {
  publicClient;
  walletClient;
  identitySDK;
  chainId;
  chainContracts;
  fallbackChains;
  rpcIterators = createRpcIteratorRegistry();
  fvDefaultChain;
  ubiSchemeAddress;
  faucetAddress;
  account;
  env;
  rdu;
  constructor({
    account,
    publicClient,
    walletClient,
    identitySDK,
    rdu = typeof window !== "undefined" ? window.location.href : "",
    env = "production"
  }) {
    if (!walletClient.account) {
      throw new Error("ClaimSDK: WalletClient must have an account attached.");
    }
    this.publicClient = publicClient;
    this.walletClient = walletClient;
    this.identitySDK = identitySDK;
    this.account = account ?? walletClient.account.address;
    this.rdu = rdu;
    this.env = env;
    const { chainId, contractEnvAddresses } = resolveChainAndContract(
      walletClient,
      env
    );
    this.chainId = chainId;
    this.chainContracts = /* @__PURE__ */ new Map([[chainId, contractEnvAddresses]]);
    const config = chainConfigs[chainId];
    this.fvDefaultChain = config.fvDefaultChain ?? chainId;
    const fallbackEntries = FALLBACK_CHAIN_PRIORITY.filter(
      (fallbackChain) => fallbackChain !== chainId
    ).map((fallbackChain) => {
      const fallbackContracts = chainConfigs[fallbackChain]?.contracts[env] ?? null;
      if (!fallbackContracts) {
        return null;
      }
      return [fallbackChain, fallbackContracts];
    }).filter(
      (entry) => entry !== null
    );
    this.fallbackChains = fallbackEntries?.map(([id]) => id);
    fallbackEntries?.forEach(([id, contracts]) => {
      this.chainContracts.set(id, contracts);
    });
    this.ubiSchemeAddress = contractEnvAddresses.ubiContract;
    this.faucetAddress = contractEnvAddresses.faucetContract;
  }
  getContractsForChain(chainId) {
    const contracts = this.chainContracts.get(chainId);
    if (!contracts) {
      throw new Error(
        `Missing contract configuration for chain ${chainId} in env ${this.env}.`
      );
    }
    return contracts;
  }
  getActiveChainId() {
    const connectedChainId = this.walletClient.chain?.id;
    if (isSupportedChain(connectedChainId)) {
      return connectedChainId;
    }
    return this.chainId;
  }
  async readChainEntitlement(chainId, client) {
    const contracts = this.getContractsForChain(chainId);
    const isPrimaryChain = chainId === this.chainId;
    const resolvedClient = client ? client : isPrimaryChain ? this.publicClient : getRpcFallbackClient(chainId, this.rpcIterators);
    let altClient;
    if (isPrimaryChain) {
      altClient = client ? resolvedClient : void 0;
    } else {
      altClient = resolvedClient;
    }
    const { root: rootAddress } = await this.identitySDK.getWhitelistedRoot(
      this.account
    );
    return this.readContract(
      {
        address: contracts.ubiContract,
        abi: ubiSchemeV2ABI,
        functionName: "checkEntitlement",
        args: [rootAddress]
      },
      altClient,
      chainId
    );
  }
  async findAltEntitlement() {
    for (const fallbackChainId of this.fallbackChains) {
      const rpcUrls = [...chainConfigs[fallbackChainId]?.rpcUrls ?? []];
      if (!rpcUrls.length) {
        continue;
      }
      for (let attempt = 0; attempt < rpcUrls.length; attempt++) {
        try {
          const amount = await this.readChainEntitlement(fallbackChainId);
          if (amount > 0n) {
            return {
              chainId: fallbackChainId,
              amount
            };
          }
        } catch {
          continue;
        }
      }
    }
    return null;
  }
  static async init(props) {
    const [account] = await props.walletClient.getAddresses();
    return new _ClaimSDK({ account, ...props });
  }
  /**
   * Reads a contract function using publicClient.
   * @param params - Parameters for the contract read operation.
   * @returns The result of the contract read.
   * @throws If the contract read fails.
   */
  async readContract(params, altClient, targetChain, attempt = 0) {
    const chainId = targetChain ?? this.chainId;
    const client = altClient || this.publicClient;
    const errorPrefix = `Failed to read contract ${params.functionName}`;
    try {
      return await client.readContract({
        address: params.address,
        abi: params.abi,
        functionName: params.functionName,
        args: params.args || [],
        account: this.account
      });
    } catch (error) {
      const combinedMessage = extractErrorMessage(error);
      if (shouldRetryRpcFallback(combinedMessage, chainId, attempt)) {
        const fallbackClient = getRpcFallbackClient(chainId, this.rpcIterators);
        return this.readContract(
          params,
          fallbackClient,
          chainId,
          attempt + 1
        );
      }
      throw new Error(`${errorPrefix}: ${combinedMessage}`);
    }
  }
  /**
   * Submits a transaction and waits for its receipt.
   * @param params - Parameters for simulating the contract call.
   * @param onHash - Optional callback to receive the transaction hash.
   * @returns The transaction receipt.
   * @throws If submission fails or no active wallet address is found.
   */
  async submitAndWait(params, onHash) {
    if (!this.account) {
      throw new Error("No active wallet address found.");
    }
    const { request } = await this.publicClient.simulateContract({
      account: this.account,
      ...params
    });
    const hash = await this.walletClient.writeContract(request);
    onHash?.(hash);
    await new Promise((res) => setTimeout(res, 5e3));
    return waitForTransactionReceipt(this.publicClient, {
      hash,
      retryDelay: 5e3
    });
  }
  /**
   * Checks if the connected user is eligible to claim UBI for the current period.
   * Returns the amount they can claim (0 if not eligible or already claimed).
   * Does not check for whitelisting status.
   * @param pClient - Optional public client scoped to an alternative chain.
   * @param chainOverride - Optional chain id to evaluate entitlement against.
   * @returns The claimable amount in the smallest unit (e.g., wei).
   * @throws If the entitlement check fails.
   */
  async checkEntitlement(options = {}) {
    const targetChain = options.chainOverride ?? this.chainId;
    const clientOverride = options.publicClient;
    const isPrimaryChain = targetChain === this.chainId;
    const amount = await this.readChainEntitlement(targetChain, clientOverride);
    if (!isPrimaryChain) {
      const hasAltAmount = amount > 0n;
      return {
        amount,
        altClaimAvailable: hasAltAmount,
        altChainId: hasAltAmount ? targetChain : null,
        altAmount: hasAltAmount ? amount : null
      };
    }
    if (amount > 0n) {
      return {
        amount,
        altClaimAvailable: false,
        altChainId: null,
        altAmount: null
      };
    }
    const altClaim = await this.findAltEntitlement();
    return {
      amount,
      altClaimAvailable: Boolean(altClaim),
      altChainId: altClaim?.chainId ?? null,
      altAmount: altClaim?.amount ?? null
    };
  }
  /**
   * Checks the wallet claim status for the connected user.
   * This method provides a single point to check if the user can claim, needs verification, or has already claimed.
   * @returns WalletClaimStatus object with status, entitlement, and optionally nextClaimTime
   * @throws If unable to check wallet status or fetch required data.
   */
  async getWalletClaimStatus() {
    const userAddress = this.account;
    const { isWhitelisted } = await this.identitySDK.getWhitelistedRoot(userAddress);
    if (!isWhitelisted) {
      return {
        status: "not_whitelisted",
        entitlement: 0n
      };
    }
    const entitlementResult = await this.checkEntitlement();
    const entitlement = entitlementResult.amount;
    if (entitlement > 0n) {
      return {
        status: "can_claim",
        entitlement
      };
    }
    const nextClaimTime = await this.nextClaimTime();
    return {
      status: "already_claimed",
      entitlement: 0n,
      nextClaimTime
    };
  }
  /**
   * Attempts to claim UBI for the connected user.
   * 1. Checks if the user is whitelisted using IdentitySDK.
   * 2. If not whitelisted, redirects to face verification and throws an error.
   * 3. If whitelisted, checks if the user can claim UBI from the pool using checkEntitlement.
   * 4. If whitelisted and can claim, checks if the user has sufficient balance.
   * 5. If the user cannot claim due to low balance, triggers a faucet request and waits.
   * 6. If whitelisted and can claim, proceeds to call the claim function on the UBIScheme contract.
   * @param txConfirm - Optional callback to confirm transactions before execution.
   * @returns The transaction receipt if the claim is successful.
   * @throws If the user is not whitelisted, not entitled to claim, balance check fails, or claim transaction fails.
   */
  async claim(txConfirm) {
    const userAddress = this.account;
    const { isWhitelisted } = await this.identitySDK.getWhitelistedRoot(userAddress);
    if (!isWhitelisted) {
      await this.fvRedirect();
      throw new Error("User requires identity verification.");
    }
    const entitlementResult = await this.checkEntitlement();
    if (entitlementResult.amount === 0n) {
      throw new Error("No UBI available to claim for this period.");
    }
    const canClaim2 = await this.checkBalanceWithRetry(txConfirm);
    if (!canClaim2) {
      throw new Error("Failed to meet balance threshold after faucet request.");
    }
    try {
      return await this.submitAndWait({
        address: this.ubiSchemeAddress,
        abi: ubiSchemeV2ABI,
        functionName: "claim",
        chain: this.walletClient.chain
      });
    } catch (error) {
      if (error instanceof ContractFunctionExecutionError) {
        throw new Error(`Claim failed: ${error.shortMessage}`);
      }
      throw new Error(`Claim failed: ${error.message}`);
    }
  }
  /**
   * Redirects the user through the face-verification flow.
   * @throws If face verification redirect fails.
   */
  async fvRedirect() {
    const fvChainId = this.fvDefaultChain ?? this.chainId;
    const fvLink = await this.identitySDK.generateFVLink(
      false,
      this.rdu,
      fvChainId
    );
    if (typeof window !== "undefined") {
      window.location.href = fvLink;
    } else {
      throw new Error(
        "Face verification redirect is only supported in browser environments."
      );
    }
  }
  /**
   * Retrieves the next available claim time for the connected user.
   * Returns epoch time (0) if the user can claim now (entitlement > 0).
   * @returns The timestamp when the user can next claim UBI, or epoch time if can claim now.
   * @throws If unable to fetch the next claim time.
   */
  async nextClaimTime() {
    const entitlementResult = await this.checkEntitlement();
    if (entitlementResult.amount > 0n) {
      return /* @__PURE__ */ new Date(0);
    }
    const [periodStart, currentDay] = await Promise.all([
      this.readContract({
        address: this.ubiSchemeAddress,
        abi: ubiSchemeV2ABI,
        functionName: "periodStart"
      }),
      this.readContract({
        address: this.ubiSchemeAddress,
        abi: ubiSchemeV2ABI,
        functionName: "currentDay"
      })
    ]);
    const periodStartMs = Number(periodStart) * 1e3;
    const startRef = new Date(periodStartMs + Number(currentDay) * DAY);
    const now = /* @__PURE__ */ new Date();
    return startRef < now ? new Date(startRef.getTime() + DAY) : startRef;
  }
  /**
   * Gets the number of claimers and total amount claimed for the current day.
   * @returns An object containing the number of claimers and total amount claimed.
   * @throws If unable to fetch daily stats.
   */
  async getDailyStats() {
    const [claimers, amount] = await this.readContract({
      address: this.ubiSchemeAddress,
      abi: ubiSchemeV2ABI,
      functionName: "getDailyStats"
    });
    return { claimers, amount };
  }
  /**
   * Triggers a faucet request to top up the user's balance.
   * @param txConfirm - Optional callback to confirm transactions before execution.
   * @throws If the faucet request fails.
   *
   * NOTE: Upgraded to contract-first flow:
   *  - Try on-chain faucet call (user signs) via `faucet.topWallet(address)`
   *  - Guard against gas>topping griefing and low native balance for publishing the tx
   *  - If on-chain path fails (or cannot sign/publish), fallback to backend `/verify/topWallet`
   *  - Throttled to at most once per hour per chain (localStorage)
   */
  async triggerFaucet(txConfirm) {
    if (txConfirm) {
      const message = "A manual transaction needs to be signed in order to claim UBI. Please confirm the transaction in your wallet to proceed with the faucet request.";
      await txConfirm(message);
    }
    const chainId = this.getActiveChainId();
    const result = await triggerFaucet({
      chainId,
      account: this.account,
      publicClient: this.publicClient,
      walletClient: this.walletClient,
      faucetAddress: this.faucetAddress,
      env: this.env,
      throttleMs: 60 * 60 * 1e3
      // 1 hour
    });
    if (result === "error") {
      throw new Error("Faucet request failed");
    }
  }
  /**
   * Fetches faucet parameters (minTopping and toppingAmount).
   * @returns Faucet configuration parameters.
   * @throws If unable to fetch faucet parameters.
   */
  async getFaucetParameters() {
    const [minTopping, toppingAmount] = await Promise.all([
      this.readContract({
        address: this.faucetAddress,
        abi: faucetABI,
        functionName: "minTopping"
      }),
      this.readContract({
        address: this.faucetAddress,
        abi: faucetABI,
        functionName: "getToppingAmount"
      })
    ]);
    return { minTopping, toppingAmount };
  }
  /**
   * Checks the user's balance with retries, triggering a faucet request if needed.
   * @param txConfirm - Optional callback to confirm transactions before execution.
   * @returns True if the balance meets the threshold, false otherwise.
   * @throws If the maximum retries are exceeded or faucet request fails.
   */
  async checkBalanceWithRetry(txConfirm) {
    const maxRetries = 5;
    const retryDelay = 5e3;
    for (let attempt = 1; attempt <= maxRetries; attempt++) {
      if (txConfirm && attempt === 1) {
        const message = "You might have to sign two transactions if you need additional gas to perform your UBI claim. ";
        await txConfirm(message);
      }
      const chainId = this.getActiveChainId();
      const result = await triggerFaucet({
        chainId,
        account: this.account,
        publicClient: this.publicClient,
        walletClient: this.walletClient,
        faucetAddress: this.faucetAddress,
        env: this.env,
        throttleMs: 60 * 60 * 1e3
        // 1 hour
      });
      if (result === "skipped") return true;
      if (result === "topped_via_contract" || result === "topped_via_api")
        return true;
      if (result === "error" && attempt < maxRetries) {
        await new Promise((resolve) => setTimeout(resolve, retryDelay));
        continue;
      }
      if (result === "error") return false;
    }
    return false;
  }
};
var ClaimCustodialSDK = class extends ClaimSDK {
  constructor(options) {
    const account = options.walletClient.account?.address;
    if (!account) {
      throw new Error("ClaimCustodialSDK: WalletClient must have an account attached.");
    }
    super({
      ...options,
      account
    });
  }
  /**
   * Override submitAndWait to handle LocalAccount signing for Celo RPC compatibility
   * Submits a transaction and waits for its receipt.
   * @param params - Parameters for simulating the contract call.
   * @param onHash - Optional callback to receive the transaction hash.
   * @returns The transaction receipt.
   * @throws If submission fails or no active wallet address is found.
   */
  async submitAndWait(params, onHash) {
    const account = this.walletClient.account;
    if (!account?.address) {
      throw new Error("No active wallet address found.");
    }
    const { request } = await this.publicClient.simulateContract({
      account: account.address,
      ...params
    });
    try {
      let hash;
      if (account && "signTransaction" in account) {
        console.log("Using LocalAccount - signing transaction locally");
        const preparedRequest = await this.walletClient.prepareTransactionRequest({
          account: account.address,
          // @ts-ignore
          to: request.to,
          // @ts-ignore
          data: request.data,
          value: request.value || 0n,
          gas: request.gas,
          gasPrice: request.gasPrice,
          maxFeePerGas: request.maxFeePerGas,
          maxPriorityFeePerGas: request.maxPriorityFeePerGas
        });
        const signedTransaction = await account.signTransaction(preparedRequest);
        hash = await this.publicClient.sendRawTransaction({
          serializedTransaction: signedTransaction
        });
      } else {
        console.log("Fallback to writeContract - may fail on Celo RPC");
        hash = await this.walletClient.writeContract(request);
      }
      onHash?.(hash);
      await new Promise((res) => setTimeout(res, 5e3));
      return waitForTransactionReceipt(this.publicClient, {
        hash,
        retryDelay: 5e3
      });
    } catch (error) {
      console.error("Transaction submission failed:", error);
      if (error.message?.includes("rpc method is not whitelisted") || error.message?.includes("eth_sendTransaction") || error.code === -32601) {
        throw new Error(
          "Transaction failed: Celo RPC does not support eth_sendTransaction. This error occurs when the LocalAccount is not properly set up. Ensure you are using privateKeyToAccount() and the account is attached to walletClient."
        );
      }
      if (error.message?.includes("insufficient funds")) {
        throw new Error("Transaction failed: Insufficient funds for gas fees.");
      }
      if (error.message?.includes("nonce")) {
        throw new Error("Transaction failed: Nonce error. Please try again.");
      }
      throw new Error(`Transaction submission failed: ${error.message}`);
    }
  }
};
var IdentityCustodialSDK = class extends IdentitySDK {
  /**
   * Override submitAndWait to handle account retrieval differently for custodial wallets
   * @param params - Parameters for simulating the contract call.
   * @param onHash - Optional callback to receive the transaction hash.
   * @returns The transaction receipt.
   * @throws If submission fails or no active wallet address is found.
   */
  async submitAndWait(params, onHash) {
    try {
      const account = this.walletClient.account?.address;
      if (!account) throw new Error("No active wallet address found.");
      const { request } = await this.publicClient.simulateContract({
        account,
        ...params
      });
      const hash = await this.walletClient.writeContract(request);
      onHash?.(hash);
      return waitForTransactionReceipt(this.publicClient, { hash });
    } catch (error) {
      console.error("submitAndWait Error:", error);
      throw new Error(`Failed to submit transaction: ${error.message}`);
    }
  }
  /**
   * Override generateFVLink to handle LocalAccount message signing for Celo RPC compatibility
   * @param popupMode - Whether to generate a popup link.
   * @param callbackUrl - The URL to callback after verification.
   * @param chainId - The blockchain network ID.
   * @returns The generated Face Verification link.
   */
  async generateFVLink(popupMode = false, callbackUrl, chainId) {
    try {
      const account = this.walletClient.account;
      if (!account?.address) throw new Error("No wallet address found.");
      const nonce = Math.floor(Date.now() / 1e3).toString();
      const fvSigMessage = FV_IDENTIFIER_MSG2.replace("<account>", account.address);
      let fvSig;
      if ("signMessage" in account) {
        fvSig = await account.signMessage({
          message: fvSigMessage
        });
      } else {
        try {
          fvSig = await this.walletClient.signMessage({
            account: account.address,
            message: fvSigMessage
          });
        } catch (rpcError) {
          throw new Error(`Message signing failed: Celo RPC doesn't support personal_sign. Use a local account instead. ${rpcError.message}`);
        }
      }
      const { identityUrl } = Envs[this.env];
      if (!identityUrl) {
        throw new Error("identityUrl is not defined in environment settings.");
      }
      if (!fvSig) {
        throw new Error("Missing signature for Face Verification.");
      }
      if (!popupMode && !callbackUrl) {
        throw new Error("Callback URL is required for redirect mode.");
      }
      const url = new URL(identityUrl);
      const params = {
        account: account.address,
        nonce,
        fvsig: fvSig,
        chain: chainId || await this.publicClient.getChainId()
      };
      if (callbackUrl) {
        params[popupMode ? "cbu" : "rdu"] = callbackUrl;
      }
      url.searchParams.append(
        "lz",
        compressToEncodedURIComponent(JSON.stringify(params))
      );
      return url.toString();
    } catch (error) {
      console.error("generateFVLink Error:", error);
      throw new Error(
        `Failed to generate Face Verification link: ${error.message}`
      );
    }
  }
};

// src/capabilities.ts
var CITIZEN_SDK_ENVIRONMENTS = [
  "production",
  "staging",
  "development"
];
var CITIZEN_SDK_SUPPORTED_CHAINS = [
  122 /* FUSE */,
  42220 /* CELO */,
  50 /* XDC */
];
var citizenSdkCapabilities = {
  sdk: "@goodsdks/citizen-sdk",
  environments: CITIZEN_SDK_ENVIRONMENTS,
  chains: CITIZEN_SDK_SUPPORTED_CHAINS,
  needs: [
    "connectedWallet",
    "readClient",
    "walletClient",
    "chainSwitching"
  ],
  reads: [
    {
      id: "whitelistStatus",
      method: "IdentitySDK.getWhitelistedRoot",
      needs: ["readClient"]
    },
    {
      id: "identityExpiry",
      method: "IdentitySDK.getIdentityExpiryData",
      needs: ["readClient"]
    },
    {
      id: "claimStatus",
      method: "ClaimSDK.getWalletClaimStatus",
      needs: ["readClient", "walletClient", "connectedWallet"]
    },
    {
      id: "claimEntitlement",
      method: "ClaimSDK.checkEntitlement",
      needs: ["readClient", "walletClient", "connectedWallet"]
    },
    {
      id: "genericClaimEntitlement",
      method: "checkGenericEntitlement",
      needs: ["readClient"]
    },
    {
      id: "dailyStats",
      method: "checkGenericDailyStats",
      needs: ["readClient"]
    }
  ],
  writes: [
    {
      id: "startVerification",
      method: "IdentitySDK.generateFVLink",
      needs: ["walletClient", "connectedWallet"],
      effects: ["signature", "externalFlow"]
    },
    {
      id: "claim",
      method: "ClaimSDK.claim",
      needs: ["walletClient", "connectedWallet"],
      effects: ["transaction", "signature", "externalFlow"]
    }
  ]
};

export { CHAIN_DECIMALS, CITIZEN_SDK_ENVIRONMENTS, CITIZEN_SDK_SUPPORTED_CHAINS, ClaimCustodialSDK, ClaimSDK, Envs, FALLBACK_CHAIN_PRIORITY, FV_IDENTIFIER_MSG2, IdentityCustodialSDK, IdentitySDK, SUPPORTED_CHAIN_IDS, SupportedChains, WALLET_LINK_SECURITY_MESSAGES, chainConfigs, checkGenericDailyStats, checkGenericEntitlement, citizenSdkCapabilities, createRpcUrlIterator, faucetABI, g$ABI, identityV2ABI, initializeIdentityContract, isSupportedChain, triggerFaucet, ubiSchemeV2ABI };
